sandbox/libs/authn/src/main/java/org/opensearch/authn/PermissionInterface.java

```diff
@@ -8,12 +8,26 @@
 
 package org.opensearch.authn;
 
+/**
+ * A generic interface for what a permission handler should implement
```

Comment:
These comments are not very useful, they do not justify why one would use this interface compared to another.

Here are some open ended questions to consider that might help inform your design.
- Why/Why not separate permissions where a permission is stored from the subject?
- What is the relationship between the actions[^1] that are being permissioned vs the resources vs the subject/principals they are assigned to?
- When you lookup permissions, subjects can represent multiple principals, when is there relationship projected within the data or dynamically by performing multiple lookups?
- How could this interface support multiple storage implementation is it flexible enough for different kinds of usages, e.g. an in-memory LRU cache vs stored in a OpenSearch index


[^1]: I called this 'name' before which is too generic, action name is more appropriate distinction - welcome to modify this

Reply:
> 

1. I feel that the permissions should be stored in association to the operating subject to prevent extremely large instances of storage for clusters with many different subjects. Unless you are attaching a database structure, you would not want some data structure containing a doubly nested mapping of Subject : (Resource: Permission List).
2. The subjects are granted permissions to perform actions on a resource. 
3. I do not know that we have a strong enough differentiation between a subject and principles at this point to be able to answer this yet. I think I would need further context on the concrete difference between the two and how we want to separate them. Maybe this is my confusion but I am not sure about this yet. 
4. The interface just defines the three required operations you would need for the the handler. So you could implement the interface as pretty much any structure with as long as there is a method for PUT, GET, DELETE of permissions. 

I hope this helps give some context to what I am thinking. These are all good questions and I will continue to keep these in mind as we go forward.